### PR TITLE
HDDS-12574. Add script to find modules by test classes

### PR DESCRIPTION
**Pull request by peterxcli**
## What changes were proposed in this pull request?

Since flaky-test-check workflow run with designated TestClass, so we can find which package the TestClass belongs to then use `-pl` args to only build necessary packages

I made a patch which shows great improvement in test running time:

Test with TestDeadNodeHandler

- Run with master CI file: https://github.com/peterxcli/ozone/actions/runs/13817211733 (takes 19mins
- Run with speedup-flaky-test-check-in-ci POC branch: https://github.com/peterxcli/ozone/actions/runs/13817907647 (takes 8mins
 

This patch also support duplicated class name across different packages, and the wildcard matching

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-12574

## How was this patch tested?

CI:
~~https://github.com/peterxcli/ozone/actions/runs/13818175368~~
~~https://github.com/peterxcli/ozone/actions/runs/13820240729~~
https://github.com/peterxcli/ozone/actions/runs/13831360576

flaky check test:
- Run with master CI file: https://github.com/peterxcli/ozone/actions/runs/13817211733 (takes 19mins
- Run with speedup-flaky-test-check-in-ci POC branch: https://github.com/peterxcli/ozone/actions/runs/13817907647 (takes 8mins

### CODE REVIEW

**Comment by peterxcli** [comment]
Trigger jira binding

File: hadoop-ozone/dev-support/checks/_lib.sh [hunk 1]
```diff
@@ -72,3 +72,63 @@ create_aws_dir() {
     popd
   fi
 }
+
+
+# Function to find project paths for a given test class
+# Returns an array of project paths in PROJECT_PATHS
+find_project_paths_for_test_class() {
```
Comment (adoroszlai) [inline hadoop-ozone/dev-support/checks/_lib.sh:79]: Changes in `_lib.sh` trigger full CI run.  I would prefer to add this function as a helper script in `dev-support/ci`, and add exclusions in `dev-support/ci/selective_ci_checks.sh`, similar to `pr_title_check`.

File: hadoop-ozone/dev-support/checks/_lib.sh [hunk 2]
```diff
@@ -72,3 +72,63 @@ create_aws_dir() {
...
+  local test_class="$1"
+  # Reset the global PROJECT_PATHS array
+  PROJECT_PATHS=()
```
Comment (adoroszlai) [inline hadoop-ozone/dev-support/checks/_lib.sh:82]: I would change the script to print results to stdout and debug messages to stderr, instead of using a global variable.

**Comment by peterxcli** [comment]
Hi @adoroszlai, please let me know if you have any concerns. Thanks!

**Review by adoroszlai — COMMENTED** [review]
Thanks a lot @peterxcli for working on this.

Finding test classes that match the pattern seems to be complex, and I think we may still be missing corner cases.

I would like to propose an alternative approach, which is limited to the flaky-test-check, but finds tests for any pattern that the test runner supports.  The idea is to trigger a single test run with the pattern, without limiting by modules, but setting very short timeout, and ignoring any test failure.  This way we let Surefire/JUnit find the tests to be run.  We can then get the names of modules where any tests were executed, and limit repeated runs to those.  (Note that this also requires the changes from #8024.)

https://github.com/adoroszlai/ozone/commit/23b834bd7a691ec1f9812c6315f99e38905ca3a4
sample run: https://github.com/adoroszlai/ozone/actions/runs/13820377923

Your approach has the benefit of also working locally, not only in GitHub workflow, so we should keep the function and refine it, but not use it in `junit.sh` just yet.

To make improvements easier and safer, we should add `bats` test cases for it (see `dev-support/ci/pr_title_check.*`).

**Review by adoroszlai — COMMENTED** [review]
Thanks @peterxcli for updating the patch and adding extensive bats test coverage.

File: dev-support/ci/selective_ci_checks.sh [hunk 3]
```diff
@@ -478,6 +479,7 @@ function get_count_misc_files() {
     start_end::group_start "Count misc. files"
     local pattern_array=(
         "^dev-support/ci/pr_title_check"
+        "^dev-support/ci/find_test_class_project.sh"
```
Comment (adoroszlai) [inline dev-support/ci/selective_ci_checks.sh:482]: > Please omit `.sh` to let most checks ignore changes in `find_test_class_project.bats`.

Sorry for not mentioning this other occurrence.

**Review by adoroszlai — APPROVED** [review]

**Comment by peterxcli** [comment]
Thanks @adoroszlai for reviewing and merging.
